pkg/daemon/daemon.go

```diff
@@ -379,39 +382,68 @@ func (dn *Daemon) restartDevicePluginPod() error {
 	dn.mu.Lock()
 	defer dn.mu.Unlock()
 	glog.V(2).Infof("restartDevicePluginPod(): try to restart device plugin pod")
-	pods, err := dn.kubeClient.CoreV1().Pods(namespace).List(metav1.ListOptions{
-		LabelSelector: "app=sriov-device-plugin",
-		FieldSelector: "spec.nodeName=" + dn.name,
-	})
-	if err != nil {
-		glog.Warningf("restartDevicePluginPod(): Failed to list device plugin pod: %s", err)
+
+	var podName string
+	if err := wait.PollImmediateUntil(3*time.Second, func() (bool, error) {
+		pods, err := dn.kubeClient.CoreV1().Pods(namespace).List(metav1.ListOptions{
+			LabelSelector: "app=sriov-device-plugin",
+			FieldSelector: "spec.nodeName=" + dn.name,
+		})
+
+		if errors.IsNotFound(err) {
+			glog.Info("restartDevicePluginPod(): device plugin pod exited")
+			return true, nil
+		}
+
+		if err != nil {
+			glog.Warningf("restartDevicePluginPod(): Failed to list device plugin pod: %s, retrying", err)
+			return false, nil
+		}
+
+		if len(pods.Items) == 0 {
```

Comment:
In which case will the condition be met?

Reply:
I was checking for len > 0 in the previous version, so I assumed it was possible to have a len == 0.
Not sure about the behaviour of `List` in case of no matches.

Reply:
In the case of len==0, this snippet returns true and it continues the next wait which delete the pod.
If the len is 0, how can it be deleted.

Reply:
Nope, if len==0, podName does not get assigned and the function does not go up to the delete because it exits before.

Let me add a podToDelete bool to make it more clear.
If you can't read it, it's pretty clear the code wasn't well written.

Reply:
ack